src/main/java/frc/robot/poseestimator/PoseEstimatorMath.java

```diff
@@ -0,0 +1,91 @@
+package frc.robot.poseestimator;
+
+import edu.wpi.first.math.Matrix;
+import edu.wpi.first.math.Nat;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.geometry.Pose2d;
+import edu.wpi.first.math.geometry.Rotation2d;
+import edu.wpi.first.math.geometry.Transform2d;
+import edu.wpi.first.math.geometry.Twist2d;
+import edu.wpi.first.math.numbers.N1;
+import edu.wpi.first.math.numbers.N3;
+import frc.robot.poseestimator.observations.OdometryObservation;
+import frc.robot.poseestimator.observations.VisionObservation;
+import java.util.Optional;
+
+public class PoseEstimatorMath {
+
+    public static Twist2d addGyroToTwistCalculations(OdometryObservation observation, Twist2d twist, Rotation2d lastGyroAngle) {
+        boolean isGyroConnected = observation.getGyroAngle() != null;
+        if (isGyroConnected) {
+            twist = updateDeltaTheta(twist, observation, lastGyroAngle);
+        }
+        return twist;
+    }
+
+    public static Twist2d updateDeltaTheta(Twist2d twist, OdometryObservation observation, Rotation2d lastGyroAngle) {
+        return new Twist2d(
+                twist.dx,
+                twist.dy,
+                observation.getGyroAngle().minus(lastGyroAngle).getRadians()
+        );
+    }
+
+    public static double[] getSquaredVisionMatrix(VisionObservation observation) {
+        int rows = observation.getStandardDeviations().getNumRows();
+        double[] r = new double[rows];
+        for (int row = 0; row < rows; ++row) {
+            double standardDeviationInRow = observation.getStandardDeviations().get(row, 0);
+            r[row] = Math.pow(standardDeviationInRow,PoseEstimatorConstants.KALMAN_EXPONENT);
+        }
+        return r;
+    }
+
+    public static Matrix<N3, N3> kalmanFilterAlgorithm(double[] squaredVisionMatrix, Matrix<N3, N1> standardDeviations) {
+        Matrix<N3, N3> visionK = new Matrix<>(Nat.N3(), Nat.N3());
```

Comment:
K?

Reply:
k for kalman

Reply:
Even if you named it visionKalman I wouldn't understand what it represents